common/include/common_defs.hpp

```diff
@@ -36,15 +37,22 @@ enum resize_factor { X1 = 0, X2, X4, X8 };
 template<typename A> using AllocChar = typename std::allocator_traits<A>::template rebind_alloc<char>;
 template<typename A> using string = std::basic_string<char, std::char_traits<char>, AllocChar<A>>;
 
-// random bit
-static std::independent_bits_engine<std::mt19937, 1, uint32_t>
-  random_bit(static_cast<uint32_t>(std::chrono::system_clock::now().time_since_epoch().count()));
+// thread-safe random bit
+inline uint32_t random_bit() 
+{
+  // add additional hash(this_thread::get_id) 
+  // to make different threads won't share the same random sequence
+  static thread_local std::independent_bits_engine<std::mt19937, 1, uint32_t>
```

Comment:
Why was this converted to a wrapper? I think this should be simpler like so:
static thread_local std::independent_bits_engine<std::mt19937, 1, uint32_t>
  random_bit(static_cast<uint32_t>(std::chrono::system_clock::now().time_since_epoch().count()
      + std::hash<std::thread::id>()(std::this_thread::get_id())));

Reply:
My consideration to use function as a wrapper, is that by using a wrapper, the whole binary can share an instance per thread, instead of "global static(meaning anonymouse in C++)" will create an instance per (thread x compilation unit), which may be too much. But this is not a big issue, I will change to follow your advice.